pubsub.go

```diff
@@ -0,0 +1,125 @@
+package lazyhttp
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/dendhi31/lazyhttp/redismaint"
+)
+
+func (httprequest *Client) Consumer() error {
+	config := redismaint.Configuration{
+		RedisURL:   httprequest.PubSubServer,
+		ContexName: "first",
+		Logger:     httprequest.Logger,
+		Handler:    httprequest.optimisticReq,
+	}
+
+	rmaint, err := redismaint.New(config)
+	if err != nil {
+		return err
+	}
+
+	term := make(chan os.Signal, 1)
+	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
+
+	go func(r *redismaint.Consumer) {
+		log.Println("lazyhttp consumer started")
+		r.Run()
+	}(rmaint)
+	//send sample schedule
+	select {
+	case <-rmaint.Err():
+		return err
+	case <-term:
+		log.Println("lazyhttp consumer stopped")
+		rmaint.Stop()
+		return nil
+	}
+}
+
+func (httprequest *Client) optimisticReq(ctx context.Context, url string, action string, payload []byte, header map[string]string, key string) (int, []byte, error) {
+	var responseBody []byte
+	var err error
+	var code int
+
+	mCtx, cancel := context.WithTimeout(context.Background(), httprequest.WaitHttp*time.Second)
+	defer cancel()
+
+	redisChan := make(chan redisChannel, 1)
+	httpChan := make(chan httpChannel, 1)
+
+	go func() {
+		httprequest.getFromRedis(mCtx, key, redisChan)
+	}()
+
+	var redisResult redisChannel
+	var httpResult httpChannel
+	select {
+	case redisTempResult := <-redisChan:
+		if (redisChannel{}) != redisTempResult {
+			redisResult = redisTempResult
+		}
+	}
+
+	if (redisResult.ErrorChan == nil) && (redisResult.ResultChan != "") {
+		return http.StatusOK, []byte(redisResult.ResultChan), nil
+	}
+
+	body := bytes.NewBuffer(payload)
+	httpRequest, err := http.NewRequest(action, url, body)
+	if err != nil {
+		return 0, responseBody, err
+	}
+	for k, v := range header {
+		httpRequest.Header.Set(k, v)
+	}
+	go func() {
+		httprequest.doRequest(mCtx, httpRequest, key, httpChan)
+	}()
+exit:
+	for {
+		select {
+		case <-mCtx.Done():
+			httprequest.Logger.Debugln("HTTP wait got timeout", int(httprequest.WaitHttp))
+			err = errors.New("context timeout HTTP")
+			break exit
+		case httpResult = <-httpChan:
+			responseBody = httpResult.ResultChan
+			code = http.StatusOK
+			err = httpResult.ErrorChan
+			break exit
+		}
+	}
+	if err != nil {
+		//publish to redis
+		reqRequirement := redismaint.RequestRequirement{
+			Url:     url,
+			Action:  action,
+			Payload: payload,
+			Header:  header,
+			Key:     key,
+		}
+		reqJson, err := json.Marshal(reqRequirement)
+		if err != nil {
+			httprequest.Logger.Debugln("Error encode json: ", err.Error())
+			return http.StatusInternalServerError, responseBody, err
```

Comment:
if the error is from http client, I believe it wont return any http status, so do we. Only return http status if they said its InternalServer error. Check the http status from the client, and return as it.